src/file-manager/fm-desktop-wayland-bg-dialog.c

```diff
@@ -0,0 +1,454 @@
+
+        gtk_css_provider_load_from_data (provider, css, -1, NULL);
+
+        g_free (css);
+
+    }
+    else
+    /*Preview a background image*/
+    {
+        GdkPixbuf  *pixbuf;
+
+        pixbuf = gdk_pixbuf_new_from_file_at_scale (filename, geometry.width / 5,
+                                                    geometry.height / 5, TRUE, NULL);
+
+        gtk_image_set_from_pixbuf (GTK_IMAGE (preview_image), pixbuf);
+
+
+        /*Clear the color preview*/
+        string = g_string_new (NULL);
+        g_string_append (string, "#caja-wayland-bg-preview {");
+
+        g_string_append (string, "background-image: none;");
+        g_string_append (string, "background-color: transparent;");
+        g_string_append (string, "}");
+
+        css = g_string_free (string, FALSE);
+        gtk_css_provider_load_from_data (provider, css, -1, NULL);
+
+        g_free (css);
+        g_object_unref (pixbuf);
+    }
+}
+
+static void
+update_primary_color (GtkWidget *colorbutton1)
+{
+    const gchar *shading_type, *primary_color_str, *secondary_color_str;
+    GdkRGBA color1;
+
+    shading_type = g_settings_get_string (background_settings,
+                                          "color-shading-type");
+
+    secondary_color_str = g_settings_get_string (background_settings,
+                                               "secondary-color");
+
+    gtk_color_chooser_get_rgba (GTK_COLOR_CHOOSER (colorbutton1), &color1);
+    primary_color_str = gdk_rgba_to_string (&color1);
+
+    update_preview (FALSE, NULL, "", shading_type,
+                    primary_color_str, secondary_color_str);
+
+    g_settings_set_string (background_settings,
+                           "primary-color", primary_color_str);
+
+    g_settings_set_string (background_settings,
+                       "picture-filename", "");
+}
+
+static void
+update_secondary_color (GtkWidget *colorbutton2)
+{
+    const gchar *shading_type, *primary_color_str, *secondary_color_str;
+    GdkRGBA color2;
+
+    shading_type = g_settings_get_string (background_settings,
+                                          "color-shading-type");
+
+    primary_color_str = g_settings_get_string (background_settings,
+                                               "primary-color");
+
+    gtk_color_chooser_get_rgba (GTK_COLOR_CHOOSER (colorbutton2), &color2);
+    secondary_color_str = gdk_rgba_to_string (&color2);
+
+    g_settings_set_string (background_settings,
+                           "secondary-color", secondary_color_str);
+
+    g_settings_set_string (background_settings,
+                       "picture-filename", "");
+
+    update_preview (FALSE, NULL, "", shading_type,
+                    primary_color_str, secondary_color_str);
+
+}
+
+static void
+update_color_background_options (GtkWidget *colorbox)
+{
+    const gchar *shading_type, *primary_color_str, *secondary_color_str;
+
+    primary_color_str = g_settings_get_string (background_settings,
```

Comment:
Just like above, memory needs to be released

Reply:
New line 208 for this:
`    g_free (primary_color_str);`